include/envoy/upstream/upstream.h

```diff
@@ -646,6 +646,51 @@ using ProtocolOptionsConfigConstSharedPtr = std::shared_ptr<const ProtocolOption
  */
 class ClusterTypedMetadataFactory : public Envoy::Config::TypedMetadataFactory {};
 
+/**
+ * Interface for connection policy subscriber. `ConnectionRequestPolicy` uses
+ * this interface as a contract between the policy implementation and policy
+ * user(subscriber).
+ */
+class ConnectionRequestPolicySubscriber {
```

Comment:
I took a super cursory pass at this and while I'm really happy to get this work into HTTP/2 I'm worried about the code overlap between the HTTP/1 multiple-connection management and HTTP/2. 

I think we might do better here if we tried to factor out / enhance the connection management we already have, and then use it in the H2 connection pool.  WDYT?

Reply:
+1 to this. This was going to be my suggestion as well once we get the code a bit more stabilized. I think there will be significant duplication. Up to you whether you want to try to do it in HTTP/2 first and then figure out the overlap or try to factor out first... (There already is a common base class you can work from.)